src/persist-types/src/arrow.rs

```diff
@@ -37,49 +38,112 @@ mod proto {
 }
 pub use proto::ProtoArrayData;
 
-impl RustType<ProtoArrayData> for arrow::array::ArrayData {
-    fn into_proto(&self) -> ProtoArrayData {
-        ProtoArrayData {
-            data_type: Some(self.data_type().into_proto()),
-            length: u64::cast_from(self.len()),
-            offset: u64::cast_from(self.offset()),
-            buffers: self.buffers().iter().map(|b| b.into_proto()).collect(),
-            children: self.child_data().iter().map(|c| c.into_proto()).collect(),
-            nulls: self.nulls().map(|n| n.inner().into_proto()),
+/// Extract the list of fields for our recursive datatypes.
+fn fields_for_type(data_type: &DataType) -> &[FieldRef] {
+    match data_type {
+        DataType::Struct(fields) => fields,
+        DataType::List(field) => std::slice::from_ref(field),
+        DataType::Map(field, _) => std::slice::from_ref(field),
+        _ => &[],
+    }
+}
+
+/// Encode the array into proto.
+/// We omit the data type if it matches the expected type, to save space.
+fn into_proto_with_type(data: &ArrayData, expected_type: Option<&DataType>) -> ProtoArrayData {
+    let data_type = match expected_type {
+        Some(expected) => {
+            debug_assert_eq!(
```

Comment:
Thoughts on making this a soft assert so it's more easily caught in tests and prod? Then when we start returning `None` turn it into a hard assert?

Reply:
I didn't want to turn this into a hard assert because it changes the big-O substantially - since equality is recursive, we'd end up checking equality many times for types deep in the tree. (Which felt a bit silly for a change that is trying to avoid doing things multiple times for those types!)

I'll at least turn this into a soft-assert-no-log and we can shake that out in the followup however!